amethyst_network/src/simulation/resource.rs

```diff
@@ -0,0 +1,252 @@
+use super::{
+    client::Client,
+    guarantees::{DeliveryGuarantee, OrderingGuarantee},
+    message::Message,
+    transport::socket::Socket,
+};
+use std::{
+    collections::{vec_deque::Drain, VecDeque},
+    net::SocketAddr,
+};
+
+/// Resource serving as the owner of the underlying socket and the queue of messages to be sent.
+pub struct NetworkSimulationResource<S: Socket> {
+    pub(crate) socket: Option<S>,
+    is_server: bool,
+    server_addr: Option<SocketAddr>,
+    clients: Vec<Client>,
+    messages: VecDeque<Message>,
+}
+
+impl<S: Socket> NetworkSimulationResource<S> {
+    /// Create a new `NetworkSimulationResource` as a client
+    pub fn new_client(server_addr: SocketAddr) -> Self {
+        Self {
+            socket: None,
+            server_addr: Some(server_addr),
+            clients: Vec::new(),
+            messages: VecDeque::new(),
+            is_server: false,
+        }
+    }
+
+    /// Create a new `NetworkSimulationResource` as a server
+    pub fn new_server() -> Self {
+        Self {
+            socket: None,
+            server_addr: None,
+            clients: Vec::new(),
+            messages: VecDeque::new(),
+            is_server: true,
+        }
+    }
+
+    /// Update and return the `NetworkSimulationResource` with a bound socket
+    pub fn with_socket(mut self, socket: S) -> Self {
```

Comment:
What happens if we call this method when there is a connection with a peer up and running?
Do we want to allow setting new sockets after the resource has been initialized? In case we do not, should we consider a builder there?

Reply:
That's a good point. We likely will want to be able to arbitrarily set and drop the socket. For example, you may wish to only have the socket connected while connected to a game server instance but not otherwise.